api/src/controllers/changes.js

```diff
@@ -1,578 +1,396 @@
+const addChangeToLongpollFeed = (feed, changeObj) => {
+  appendChange(feed.results, changeObj);
+  // check if this newly added change would update the auth state on any of the previous denied changes
+  feed.limit -= processPendingChanges(feed, feed.results, false).nbrAppended + 1;
+
+  if (feed.limit) {
+    // debounce sending results if the feed limit is not yet reached
+    addTimeout(feed);
+    feed.debounceEnd();
     return;
   }
 
-  var modifiedChanges = changes.results.map(function(change) {
-    var result = {
-      id: change.id,
-      doc: change.doc
+  endFeed(feed);
+};
+
+const processChange = (change, seq) => {
+  let changeObj;
+  if (tombstoneUtils.isTombstoneId(change.id)) {
+    changeObj = {
+      change: tombstoneUtils.generateChangeFromTombstone(change),
+      viewResults: authorization.getViewResults(tombstoneUtils.extractDoc(change.doc))
     };
-    var row = getReplicationKey(change.doc);
-    if (row && row.length) {
-      result.subject = row[0];
-      result.submitter = row[1].submitter;
-    }
-    return result;
+  } else {
+    changeObj = {
+      change: change,
+      viewResults: authorization.getViewResults(change.doc)
+    };
+  }
+  // inform the normal feeds that a change was received while they were processing
+  normalFeeds.forEach(feed => {
+    feed.lastSeq = seq;
+    feed.pendingChanges.push(changeObj);
   });
-  continuousFeeds.forEach(function(feed) {
-    // check if new and relevant
-    if (hasNewApplicableDoc(feed, modifiedChanges)) {
-      abortUpstreamBatches(feed, 'newer-change-received', 'INFO [onRelevantChange] cancelling upstream batch (before bindServerIds()):');
-
-      bindServerIds(feed, function(err) {
-        abortUpstreamBatches(feed, 'newer-change-received', 'INFO [onRelevantChange] cancelling upstream batch (after bindServerIds()):');
-        if (err) {
-          return serverUtils.error(err, feed.req, feed.res);
-        }
-        feed.log('updateFeeds()', 'Calling getChanges()…');
-        getChanges(feed);
-      });
+
+  // send the change through to the longpoll feeds which are allowed to see it
+  longpollFeeds.forEach(feed => {
+    feed.lastSeq = seq;
+    if (!authorization.allowedChange(feed, changeObj)) {
+      feed.pendingChanges.push(changeObj);
+      return;
+    }
+
+    if (authorization.isAuthChange(feed, changeObj)) {
+      endFeed(feed, false);
+      processRequest(feed.req, feed.res, feed.userCtx);
+      return;
     }
+
+    addChangeToLongpollFeed(feed, changeObj);
   });
 };
 
-var init = function(since) {
+const initContinuousFeed = since => {
+  continuousFeed = db.medic
+    .changes({
+      live: true,
+      include_docs: true,
+      since: since || 'now',
+      timeout: false,
+    })
+    .on('change', (change, pending, seq) => {
+      processChange(change, seq);
+      currentSeq = seq;
+    })
+    .on('error', () => {
+      continuousFeed.cancel();
+      Promise.resolve().then(() => {
```

Comment:
I don't think this needs to be wrapped in a `Promise.resolve` - are you using it as a cooler version of `setTimeout`?